File: app/src/main/java/com/ddancn/keepaccount/activity/MainActivity.java
package com.ddancn.keepaccount.activity;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentPagerAdapter;
import androidx.viewpager.widget.ViewPager;

import com.ddancn.keepaccount.R;
import com.ddancn.keepaccount.fragment.AddFragment;
import com.ddancn.keepaccount.fragment.RecordFragment;
import com.ddancn.keepaccount.fragment.SumFragment;
import com.ddancn.lib.base.BaseActivity;
import com.google.android.material.bottomnavigation.BottomNavigationView;

import java.util.ArrayList;
import java.util.List;

/**
 * @author ddan.zhuang
 */
public class MainActivity extends BaseActivity {

    private ViewPager mViewPager;
    private BottomNavigationView mBottomNavView;

    @Override
    protected int bindLayout() {
        return R.layout.activity_main;
    }

    @Override
    protected boolean hasHeader() {
        return false;
    }

    @Override
    protected void initView() {
        mViewPager = findViewById(R.id.viewpager);
        mBottomNavView = findViewById(R.id.navigation);
        MainPagerAdapter adapter = new MainPagerAdapter(getSupportFragmentManager());
        adapter.addFragment(new AddFragment());
        adapter.addFragment(new RecordFragment());
        adapter.addFragment(new SumFragment());
        mViewPager.setAdapter(adapter);
    }

    @Override
    protected void bindListener() {
        mBottomNavView.setOnNavigationItemSelectedListener(item -> {
            switch (item.getItemId()) {
                case R.id.navigation_add:
                    mViewPager.setCurrentItem(0);
                    return true;
                case R.id.navigation_list:
                    mViewPager.setCurrentItem(1);
                    return true;
                case R.id.navigation_chart:
                    mViewPager.setCurrentItem(2);
                    return true;
                default:
            }
            return false;
        });
        mViewPager.addOnPageChangeListener(new ViewPager.SimpleOnPageChangeListener() {
            @Override
            public void onPageSelected(int position) {
                mBottomNavView.getMenu().getItem(position).setChecked(true);
            }
        });
    }

    public class MainPagerAdapter extends FragmentPagerAdapter {

        private List<Fragment> fragments = new ArrayList<>();

        public MainPagerAdapter(FragmentManager fm) {
            super(fm, BEHAVIOR_RESUME_ONLY_CURRENT_FRAGMENT);
        }

        @Override
        @NonNull
        public Fragment getItem(int position) {
            return fragments.get(position);
        }

        @Override
        public int getCount() {
            return fragments.size();
        }

        public void addFragment(Fragment fragment) {
            fragments.add(fragment);
        }
    }
}

File: app/src/main/java/com/ddancn/keepaccount/activity/MainActivity.kt
package com.ddancn.keepaccount.activity

import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.viewpager2.adapter.FragmentStateAdapter
import androidx.viewpager2.widget.ViewPager2
import com.ddancn.keepaccount.R
import com.ddancn.keepaccount.databinding.ActivityMainBinding
import com.ddancn.keepaccount.fragment.AddFragment
import com.ddancn.keepaccount.fragment.RecordFragment
import com.ddancn.keepaccount.fragment.SumFragment
import com.ddancn.lib.base.BaseActivity

/**
 * @author ddan.zhuang
 */
class MainActivity : BaseActivity<ActivityMainBinding>() {

    override fun hasHeader(): Boolean {
        return false
    }

    override fun initView() {
        val adapter = MainPagerAdapter(this)
        adapter.addFragment(AddFragment())
        adapter.addFragment(RecordFragment())
        adapter.addFragment(SumFragment())
        vb.viewpager.adapter = adapter
        vb.viewpager.offscreenPageLimit = 2
    }

    override fun bindListener() {
        vb.navigation.setOnItemSelectedListener { item ->
            when (item.itemId) {
                R.id.navigation_add -> vb.viewpager.currentItem = 0
                R.id.navigation_list -> vb.viewpager.currentItem = 1
                R.id.navigation_chart -> vb.viewpager.currentItem = 2
            }
            false
        }
        vb.viewpager.registerOnPageChangeCallback(object : ViewPager2.OnPageChangeCallback() {
            override fun onPageSelected(position: Int) {
                vb.navigation.menu.getItem(position).isChecked = true
            }
        })
    }

    class MainPagerAdapter(activity: FragmentActivity) : FragmentStateAdapter(activity) {

        private val fragments = ArrayList<Fragment>()

        fun addFragment(fragment: Fragment) {
            fragments.add(fragment)
        }

        override fun getItemCount(): Int {
            return fragments.size
        }

        override fun createFragment(position: Int): Fragment {
            return fragments[position]
        }
    }
}
